inlong-sort-standalone/sort-standalone-source/src/main/java/org/apache/inlong/sort/standalone/source/sortsdk/FetchCallback.java

```diff
@@ -0,0 +1,138 @@
+ *            doSomething();
+ *            final ProfileEvent profileEvent = new ProfileEvent(result.getBody(), result.getHeaders(), acker);
+ *             channelProcessor.processEvent(profileEvent);
+ *        }
+ *
+ * The ACKer will be used to <b>ACK</b> upstream after that the downstream <b>ACKed</b> sort-standalone.
+ * This process seems like <b>transaction</b> of the whole sort-standalone, and which
+ * ensure <b>At Least One</b> semantics.
+ */
+public class FetchCallback implements ReadCallback {
+
+    // Logger of {@link FetchCallback}.
+    private static final Logger LOG = LoggerFactory.getLogger(FetchCallback.class);
+
+    // SortId of fetch message.
+    private final String sortId;
+
+    // ChannelProcessor that put message in specific channel.
+    private final ChannelProcessor channelProcessor;
+
+    // Context of source, used to report fetch results.
+    private final SortSdkSourceContext context;
+
+    // Temporary usage for ACK. The {@link SortClient} and Callback should not circular reference each other.
+    private SortClient client;
+
+    /**
+     * Private constructor of {@link FetchCallback}.
+     * <p> The construction of FetchCallback should be initiated by {@link FetchCallback.Factory}.</p>
+     *
+     * @param sortId SortId of fetch message.
+     * @param channelProcessor ChannelProcessor that message put in.
+     * @param context The context to report fetch results.
+     */
+    private FetchCallback(
+            final String sortId,
+            final ChannelProcessor channelProcessor,
+            final SortSdkSourceContext context) {
+        this.sortId = sortId;
+        this.channelProcessor = channelProcessor;
+        this.context = context;
+    }
+
+    /**
+     * Set client for ack.
+     * @param client client for ack.
+     */
+    public void setClient(@NotNull SortClient client) {
+        this.client = client;
+    }
+
+    /**
+     * The callback function that SortSDK invoke when fetch messages.
+     *
+     * <p> In order to ACK the fetched msg, {@link FetchCallback} has to hold the {@link SortClient} which results in
+     * <b>Cycle Reference</b>. The {@link SortClient} should deliver one object to ACK after invoke the callback method
+     * {@link ReadCallback#onFinished(MessageRecord)}. </p>
+     *
+     * @param messageRecord message
+     */
+    @Override
+    public void onFinished(final MessageRecord messageRecord) {
+        try {
+            Preconditions.checkState(messageRecord != null, "Fetched msg is null.");
+            final SubscribeFetchResult result = SubscribeFetchResult.Factory.create(sortId, messageRecord);
+            final ProfileEvent profileEvent = new ProfileEvent(result.getBody(), result.getHeaders());
+            channelProcessor.processEvent(profileEvent);
+            context.reportToMetric(profileEvent, sortId, "-", SortSdkSourceContext.FetchResult.SUCCESS);
+            client.ack(messageRecord.getMsgKey(), messageRecord.getMsgKey());
```

Comment:
The acking operation need to move to ProfileEvent. 
In the future, execute acking operation after sending successfully.

Reply:
As the explanation in javadoc of FetchCallback.java and callback method "onFinished", the ACKing operation is done by SortSdk client, which means that the FetchCallback class must held one instance of corresponding client while the client hold callback object.  In order to avoid cycle reference, the SDK client should pass though one object for acking, and this object will be held by ProfileEvent and used to ack after receiving the ACK from downstream. The current acking operation is just a temporary usage.